File: packages/platform-mobile/android/app/src/main/java/app/bramble/mobile/AutofillCaps.kt
package app.bramble.mobile

import android.content.Context
import org.json.JSONObject
import java.io.File

// Records device autofill capabilities the app can't otherwise probe, so Settings can hide
// controls that wouldn't work here. Inline-suggestion support (iOS-QuickType equivalent) is
// only revealed to the service during a fill request (request.inlineSuggestionsRequest) and
// depends on the active keyboard: Gboard / SwiftKey support it, the AOSP keyboard does not.
// Stored as a file (not SharedPreferences) so the main app process reads a fresh value across
// the :autofill process boundary without the multi-process SharedPreferences cache hazard.
object AutofillCaps {
    private const val FILE = "autofill_caps.json"

    fun recordInlineSupport(context: Context, supported: Boolean) {
        val f = File(context.filesDir, FILE)
        val current = runCatching { JSONObject(f.readText()).optBoolean("inlineSupported") }.getOrNull()
        if (current == supported) return
        runCatching { f.writeText(JSONObject().put("inlineSupported", supported).toString()) }
    }
}

File: packages/platform-mobile/src/autofill-caps.ts
import { Directory, Encoding, Filesystem } from "@capacitor/filesystem";

// Whether the device's active keyboard can show inline autofill suggestions (the Android
// equivalent of iOS QuickType). The native AutofillService records this on each fill request
// (it depends on the keyboard: Gboard / SwiftKey yes, the AOSP keyboard no). It is unknown
// until the user has autofilled once, so default to false: Settings hides the
// keyboard-suggestions control until the capability is confirmed, rather than show a control
// that can't work. See docs/mobile-port.md.
const FILE = "autofill_caps.json";
const DIR = Directory.Data;

export async function inlineSuggestionsSupported(): Promise<boolean> {
	try {
		const r = await Filesystem.readFile({ path: FILE, directory: DIR, encoding: Encoding.UTF8 });
		return JSON.parse(r.data as string)?.inlineSupported === true;
	} catch {
		return false;
	}
}
